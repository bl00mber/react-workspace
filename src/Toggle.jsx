class Toggle extends React.Component {
  constructor(props) {
    super(props);
    this.state = { isToggleOn: false };
  }

  handleToggle(e, toggleValue) {
    this.setState(prevState => ({
      isToggleOn: toggleValue !== undefined ? toggleValue : !prevState.isToggleOn
    }));
  }

  render() {
    return (
      <div>
        <p>Is toggled {this.state.isToggleOn ? 'ON' : 'OFF'}</p>
        <button onClick={(e) => this.handleToggle(e)}>Toggle</button>
        <button onClick={(e) => this.handleToggle(e, true)}>Toggle ON</button>
        <button onClick={(e) => this.handleToggle(e, false)}>Toggle OFF</button>
      </div>
    );
  }
}
