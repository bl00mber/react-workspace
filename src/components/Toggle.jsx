import React from 'react';
import PropTypes from 'prop-types';

const Toggle = ({ isToggleOn, onToggle }) => (
  <div>
    <p>
      Is toggled
      {isToggleOn ? 'ON' : 'OFF'}
    </p>
    <button type="button" onClick={() => onToggle()}>Toggle</button>
    <button type="button" onClick={() => onToggle(true)}>Toggle ON</button>
    <button type="button" onClick={() => onToggle(false)}>Toggle OFF</button>
  </div>
);

Toggle.propTypes = {
  isToggleOn: PropTypes.bool.isRequired,
  onToggle: PropTypes.func.isRequired,
};

export default Toggle;
