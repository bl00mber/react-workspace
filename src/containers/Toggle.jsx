import { connect } from 'react-redux';

import ToggleComponent from '@/components/Toggle';
import { toggleFlag } from '@/store/actions';

const mapStateToProps = state => ({
  isToggleOn: state.toggle,
});

const mapDispatchToProps = dispatch => ({
  onToggle: flag => dispatch(toggleFlag(flag)),
});

const Toggle = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ToggleComponent);

export default Toggle;
