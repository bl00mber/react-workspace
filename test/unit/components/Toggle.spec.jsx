import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Toggle from '@/components/Toggle';

describe('Toggle', () => {
  test('should match react-test-renderer snapshot', () => {
    const props = {
      isToggleOn: false,
      onToggle: jest.fn(() => {
        props.isToggleOn = !props.isToggleOn;
      }),
    };
    let component = renderer.create(<Toggle {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // manually trigger the toggle
    tree.children[1].props.onClick();
    expect(props.onToggle).toHaveBeenCalledWith();

    // re-rendering
    component = renderer.create(<Toggle {...props} />);
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // manually trigger the toggle
    tree.children[1].props.onClick();
    expect(props.onToggle).toHaveBeenCalledWith();
    // re-rendering
    component = renderer.create(<Toggle {...props} />);
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should match enzyme shallow snapshot', () => {
    const props = {
      isToggleOn: false,
      onToggle: jest.fn(() => {
        props.isToggleOn = !props.isToggleOn;
      }),
    };
    const wrapper = shallow(<Toggle {...props} />);
    expect(wrapper.find('button')).toHaveLength(3);

    expect(toJson(wrapper)).toMatchSnapshot();

    // manually trigger the toggle
    wrapper.find('button').at(0).simulate('click');
    expect(props.onToggle).toHaveBeenCalledWith();
    wrapper.setProps(props);
    expect(toJson(wrapper)).toMatchSnapshot();

    // manually trigger the toggle
    wrapper.find('button').at(0).simulate('click');
    expect(props.onToggle).toHaveBeenCalledWith();
    wrapper.setProps(props);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should match enzyme mount snapshot', () => {
    const props = {
      isToggleOn: false,
      onToggle: jest.fn(() => {
        props.isToggleOn = !props.isToggleOn;
      }),
    };
    const wrapper = mount(<Toggle {...props} />);
    expect(wrapper.find('button')).toHaveLength(3);

    expect(toJson(wrapper)).toMatchSnapshot();

    // manually trigger the toggle
    wrapper.find('button').at(0).simulate('click');
    expect(props.onToggle).toHaveBeenCalledWith();
    wrapper.setProps(props);
    expect(toJson(wrapper)).toMatchSnapshot();

    // manually trigger the toggle
    wrapper.find('button').at(0).simulate('click');
    expect(props.onToggle).toHaveBeenCalledWith();
    wrapper.setProps(props);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
