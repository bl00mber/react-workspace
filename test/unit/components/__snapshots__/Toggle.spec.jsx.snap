// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should match enzyme mount snapshot 1`] = `
<Toggle
  isToggleOn={false}
  onToggle={[MockFunction]}
>
  <div>
    <p>
      Is toggled
      OFF
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Toggle
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Toggle ON
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Toggle OFF
    </button>
  </div>
</Toggle>
`;

exports[`Toggle should match enzyme mount snapshot 2`] = `
<Toggle
  isToggleOn={true}
  onToggle={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
>
  <div>
    <p>
      Is toggled
      ON
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Toggle
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Toggle ON
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Toggle OFF
    </button>
  </div>
</Toggle>
`;

exports[`Toggle should match enzyme mount snapshot 3`] = `
<Toggle
  isToggleOn={false}
  onToggle={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
>
  <div>
    <p>
      Is toggled
      OFF
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Toggle
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Toggle ON
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Toggle OFF
    </button>
  </div>
</Toggle>
`;

exports[`Toggle should match enzyme shallow snapshot 1`] = `
<div>
  <p>
    Is toggled
    OFF
  </p>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle ON
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle OFF
  </button>
</div>
`;

exports[`Toggle should match enzyme shallow snapshot 2`] = `
<div>
  <p>
    Is toggled
    ON
  </p>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle ON
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle OFF
  </button>
</div>
`;

exports[`Toggle should match enzyme shallow snapshot 3`] = `
<div>
  <p>
    Is toggled
    OFF
  </p>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle ON
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle OFF
  </button>
</div>
`;

exports[`Toggle should match react-test-renderer snapshot 1`] = `
<div>
  <p>
    Is toggled
    OFF
  </p>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle ON
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle OFF
  </button>
</div>
`;

exports[`Toggle should match react-test-renderer snapshot 2`] = `
<div>
  <p>
    Is toggled
    ON
  </p>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle ON
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle OFF
  </button>
</div>
`;

exports[`Toggle should match react-test-renderer snapshot 3`] = `
<div>
  <p>
    Is toggled
    OFF
  </p>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle ON
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Toggle OFF
  </button>
</div>
`;
